src/Features/CSharp/Portable/ConvertLinq/CSharpConvertLinqQueryToForEachProvider.cs

```diff
@@ -0,0 +1,986 @@
+                    if (extraStatement != null)
+                    {
+                        statements.Add(extraStatement);
+                    }
+                }
+
+                // The stack was processed in the reverse order, but the extra statements should be provided in the direct order.
+                statements.Reverse();
+                statements.Add(statement.WithAdditionalAnnotations(Simplifier.Annotation));
+                return statements.ToArray();
+            }
+
+            private bool TryProcessQueryBody(QueryBodySyntax queryBody, QueryExpressionProcessingInfo queryExpressionProcessingInfo)
+            {
+                do
+                {
+                    foreach (var queryClause in queryBody.Clauses)
+                    {
+                        switch (queryClause.Kind())
+                        {
+                            case SyntaxKind.WhereClause:
+                                queryExpressionProcessingInfo.Add(queryClause);
+                                break;
+                            case SyntaxKind.LetClause:
+                                if (!queryExpressionProcessingInfo.TryAdd(queryClause, ((LetClauseSyntax)queryClause).Identifier))
+                                {
+                                    return false;
+                                }
+
+                                break;
+                            case SyntaxKind.FromClause:
+                                var fromClause = (FromClauseSyntax)queryClause;
+                                if (!queryExpressionProcessingInfo.TryAdd(queryClause, fromClause.Identifier))
+                                {
+                                    return false;
+                                }
+
+                                break;
+                            case SyntaxKind.JoinClause:
+                                var joinClause = (JoinClauseSyntax)queryClause;
+                                if (joinClause.Into == null) // GroupJoin is not supported
+                                {
+                                    if (queryExpressionProcessingInfo.TryAdd(joinClause, joinClause.Identifier))
+                                    {
+                                        break;
+                                    }
+                                }
+
+                                return false;
+                            // OrderBy is not supported by foreach.
+                            default:
+                                return false;
+                        }
+                    }
+
+                    // GroupClause is not supported by the conversion
+                    if (!(queryBody.SelectOrGroup is SelectClauseSyntax selectClause))
+                    {
+                        return false;
+                    }
+
+                    if (_semanticModel.GetTypeInfo(selectClause.Expression).Type.ContainsAnonymousType())
```

Comment:
>if (_semanticModel.GetTypeInfo(selectClause.Expression).Type.ContainsAnonymousType()) [](start = 20, length = 85)

This feels too conservative, but we can discuss this and, if necessary, follow-up later. Just put this on the list

Reply:
Added to [https://github.com/dotnet/roslyn/issues/25639](https://github.com/dotnet/roslyn/issues/25639)

---
In reply to: [178423888](https://github.com/dotnet/roslyn/pull/25362#discussion_r178423888) [](ancestors = 178423888)